README.md
# Google Fonts Webpack Plugin

Download Google fonts to webpack build folder using [google-webfonts-helper](https://github.com/majodev/google-webfonts-helper) or use CDN to integrate with [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin).

---

## Install

```bash
npm install google-fonts-webpack-plugin
```

## Usage

`/webpack.config.js`
```js
const GoogleFontsPlugin = require("google-fonts-webpack-plugin")

module.exports = {
	"entry": "index.js",
	/* ... */
	plugins: [
		new GoogleFontsPlugin({
			fonts: [
				{ family: "Source Sans Pro" },
				{ family: "Roboto", variants: [ "400", "700italic" ] }
			]
			/* ...options */
		})
	]
}
```

## Options

```js
new GoogleFontsPlugin(options: Object)
```

|Name|Type|Default|Description|
|----|----|-------|-----------|
|fonts|`FontObject[]`|-|Configuration generated by [http://fontello.com](Fontello.com).
|name|`String`|`"fonts"`|Module name.
|filename|`String`|`"fonts.css"`|Css file name.
|path|`String`|`"font/"`|Relative path to fonts directory. If path is `undefined` fonts are not downloaded, the css file is generated with Google hosted font files.
|local|`Boolean`|`true`|Wether to use google-webfonts-helper API or just link to Google Fonts hosted css. If this option is set to false this plugin just adds the css url to html-webpack-plugin (if present).
|formats|`Array`|`[ "eot", "woff", "woff2", "ttf", "svg" ]`|Font formats to download.
|apiUrl|`String`|`"https://google-webfonts-helper.herokuapp.com/api/fonts"`|google-webfonts-helper API url.


### FontObject

|Name|Type|Default|Description|
|----|----|-------|-----------|
|family|`String`|-|Font family.
|variants|`Array`|-|Font variants according	to google-webfonts-helper (e.g.: `[ "italic", "500", "700italic" ]`).
|subsets|`Array`|-|Font subsets according	to google-webfonts-helper (e.g.: `[ "latin", "greek" ]`).
|formats|`Array`|-|Font formats to download. Defaults to `options.formats`.


CHANGELOG.md
# Changelog

## [Unreleased]

## [v0.4.4] - 2017-12-22

### Fixed
* Fixed local public path [PR [\#16](https://github.com/gabiseabra/google-fonts-webpack-plugin/pull/16)] [issue [\#15](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/15)]

## [v0.4.3] - 2017-09-20

### Fixed
* Fixed googleapis url generator [issue [\#13](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/13)]
* Fixed README [PR [\#8](https://github.com/gabiseabra/google-fonts-webpack-plugin/pull/8)]

## [v0.4.2] - 2017-08-25

### Fixed
* Fixed peerDependencies webpack version [issue [\#12](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/12)]

## [v0.4.1] - 2017-07-3

### Fixed

* Fixed inconsistent ordering of selectors in generated css file [issue [\#6](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/6)]
* Fixed assets paths on windows

### Added

* This CHANGELOG.md [issue [\#5](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/5)]

## [v0.4.0] - 2017-06-3

### Fixed

* Fixed assets path in html-webpack-plugin [pull request [\#4](https://github.com/gabiseabra/google-fonts-webpack-plugin/pull/4)]
* Fixed relative paths in generated css files [issue [\#3](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/3)]

## [v0.3.2] - 2017-05-18

### Changed

* Switched dependencies from node-zip2 to yauzl [issue [\#1](https://github.com/gabiseabra/google-fonts-webpack-plugin/issues/1)]

## [v0.3.1] - 2017-04-22

### Changed

* Improved error handling

## [v0.3.0] - 2017-04-13

### Added

* Added README.md

### Fixed

* Fixed google fonts url generator

## [v0.2.0] - 2017-03-25

### Fixed

* Fixed error with html-webpack-plugin

## [v0.1.0] - 2017-03-24

* Initial release


spec/GoogleWebfonts.js
const should = require("should")
const GoogleWebfonts = require("../src/GoogleWebfonts")

describe("GoogleWebfonts", () => {
	let api

	before(() => {
		api = new GoogleWebfonts()
	})

	describe("#getFonts()", () => {
		it("returns a list of fonts", () => (
			api.getFonts()
				.then(fonts => {
					fonts.should.be.Array()
					.and.should.not.be.empty
				})
		))
	})

	describe("#getFontById(id)", () => {
		it("returns a font by id", () => (
			api.getFontById("lato")
				.then(font => {
					font.should.be.instanceOf(GoogleWebfonts.Font)
				})
		))
	})

	describe("#getFontByFamily(family)", () => {
		it("returns a font by family", () => (
			api.getFontByFamily("Open Sans")
				.then(font => {
					font.should.be.instanceOf(GoogleWebfonts.Font)
				})
		))
	})

	describe("Font", () => {
		let promise

		before(() => {
			promise = api.getFontById("open-sans")
		})

		describe("#getName(variant?)", () => {
			it("returns the font's file name", () => (
				promise.then(font => {
						font.getName().should.equal("OpenSans")
						font.getName("700").should.equal("OpenSans-Bold")
					})
			))
		})

		describe("#info", () => {
			it("returns the full font info", () => (
					promise.then(font => font.info())
						.then(info => {
							info.should.be.an.Object()
						})
			))
		})

		describe("#select", () => {
			it("returns a downloader", () => (
					promise.then(font => {
						font.select().should.be.instanceOf(GoogleWebfonts.Selection)
					})
			))
		})
	})

	describe("Selection", () => {
		let promise

		before(() => {
			promise = api.getFontById("open-sans")
				.then(font => font.select({
					variants: [ "400", "700italic" ],
					formats: [ "woff", "woff2" ]
				}))
		})

		describe("#css", () => {
			it("returns font-face css", () => (
				promise.then(query => query.css())
					.then(css => {
						css.should.containEql("@font-face")
					})
			))
		})

		describe("#download", () => {
			it("fetches font files", () => (
				promise.then(query => query.download())
					.then(res => {
						res.status.should.equal(200)
					})
			))
		})

		describe("#files", () => {
			it("returns a list of font files", () => (
				promise.then(query => query.files())
					.then(files => {
						files.should.have.keys(
							"OpenSans-Regular.woff",
							"OpenSans-Regular.woff2",
							"OpenSans-BoldItalic.woff",
							"OpenSans-BoldItalic.woff2"
						).and.should.not.have.keys(
							"OpenSans-Regular.eot",
							"OpenSans-Regular.ttf"
						)
					})
			))
		})

		describe("#assets", () => {
			it("returns a list of font files", () => (
				promise.then(query => query.assets())
					.then(assets => {
						assets.should.have.keys(
							"OpenSans-Regular.woff",
							"OpenSans-Regular.woff2",
							"OpenSans-BoldItalic.woff",
							"OpenSans-BoldItalic.woff2"
						).and.should.not.have.keys(
							"OpenSans-Regular.eot",
							"OpenSans-Regular.ttf"
						)
					})
			))
		})
	})
})


spec/GoogleWebfontsPlugin.js
const should = require("should")
const webpack = require("webpack")
const MemoryFs = require("memory-fs")
const GoogleWebfontPlugin = require("../src")

const webpackConfig = {
	entry: require.resolve("../package.json"),
	output: {
		path: "/",
		filename: "[name].js"
	},
	stats: {
		colors: true
	}
}

describe("GoogleWebfontPlugin", () => {
	let fs, plugin, stats

	before(() => {
		fs = new MemoryFs()
		plugin = new GoogleWebfontPlugin({
			filename: "styles/fonts.css",
			fonts: [
				{
					family: "Source Sans Pro",
					variants: [ "400", "700italic" ]
				},
				{
					family: "Roboto"
				}
			]
		})
		const compiler = webpack(Object.assign({}, webpackConfig, {
			plugins: [ plugin ]
		}))
		compiler.outputFileSystem = fs
		return new Promise((resolve, reject) => {
			compiler.run((err, data) => {
				if(err) {
					reject(err)
				} else {
					stats = data
					console.log(stats.toString(webpackConfig.stats))
					console.log("\n")
					resolve(stats)
				}
			})
		})
	})

	it("adds a chunk to compilation", () => {
		stats.compilation.namedChunks
		.should.have.key(plugin.options.name)
	})

	it("emits font files", () => {
		fs.existsSync("/font/SourceSansPro-Regular.woff").should.be.ok()
		fs.existsSync("/font/Roboto-Regular.woff").should.be.ok()
	})

	it("emits a css file", () => {
		fs.existsSync("/styles/fonts.css").should.be.ok()
		fs.readFileSync("/styles/fonts.css", "utf8")
		.should.containEql("../font/Roboto-Regular.woff")
	})
})


src/index.js
const _ = require("lodash")
const path = require("path")
const Chunk = require("webpack/lib/Chunk")
const { RawSource } = require("webpack-sources")
const GoogleWebfonts = require("./GoogleWebfonts")
const cssUrl = require("./cssUrl")

const defaults = {
	fonts: undefined,
	name: "fonts",
	apiUrl: undefined,
	formats: undefined,
	filename: "fonts.css",
	path: "font/",
	local: true
}

class GoogleWebfontsPlugin {
	constructor(options) {
		this.options = Object.assign({}, defaults, options)
		this.chunk = new Chunk(this.options.name)
		this.chunk.ids = []
		this.chunk.name = this.options.name
	}

	get api() {
		if(!this._api) {
			this._api = new GoogleWebfonts(this.options.apiUrl)
		}
		return this._api
	}

	fetch() {
		const {
			fonts,
			apiUrl,
			path: fontsPath,
			filename: cssFile,
			formats: defaultFormats
		} = this.options
		const compareCss = (a, b) => (a.id.localeCompare(b.id))
		const fontsCss = []
		const files = {}
		const promises = []
		fonts.forEach((fontOptions) => {
			const { family } = fontOptions
			const query = this.api.getFontByFamily(family)
				.then(font => {
					if(!font) {
						throw new Error(`Font family \"${family}\" not found.`)
					}
					return font.select(_.assign(
						fontOptions,
						{ formats: defaultFormats }
					))
				})
			const cssRelativePath = path.posix.relative(
				path.dirname(cssFile),
				fontsPath
			)
			promises.push(
				query.then(selection => {
					return selection.css(cssRelativePath).then((css) => {
						fontsCss.push({
							css,
							id: selection.font.id
						})
					})
				})
			)
			if(fontsPath) {
				promises.push(
					query.then(q => q.assets())
					.then(assets => {
						for(const fileName in assets) {
							files[fontsPath + fileName] = assets[fileName]
						}
					})
				)
			}
		})
		return Promise.all(promises)
			.then(() => {
				fontsCss.sort(compareCss)
				const css = fontsCss.map(font => font.css).join("\n")
				return { css, files }
			})
	}

	apply(compiler) {
		const { fonts, local, filename: cssFile } = this.options
		compiler.plugin("make", (compilation, cb) => {
			if(local) {
				const addFile = (fileName, source) => {
					this.chunk.files.push(fileName)
					compilation.assets[fileName] = source
				}
				this.fetch().then(({ css, files }) => {
					addFile(cssFile, new RawSource(css))
					for(const fileName in files) {
						addFile(fileName, files[fileName])
					}
					cb()
				})
			} else {
				cb()
			}
			compilation.plugin("html-webpack-plugin-before-html-generation", (data, cb) => {
				if (local && (data.assets.publicPath.indexOf("://") !== -1 || data.assets.publicPath.indexOf(":") !== -1)) {
                    data.assets.css.push(data.assets.publicPath + cssFile);
                } else if (local) {
                    data.assets.css.push(path.posix.join(data.assets.publicPath, cssFile));
                } else {
                    data.assets.css.push(cssUrl(fonts));
                }
				cb(null, data)
			})
			compilation.plugin("additional-assets", cb => {
				compilation.chunks.push(this.chunk)
				compilation.namedChunks[this.options.name] = this.chunk
				cb()
			})
		})
	}
}

GoogleWebfontsPlugin.GoogleWebfonts = GoogleWebfonts

GoogleWebfontsPlugin.cssUrl = cssUrl

module.exports = GoogleWebfontsPlugin


src/GoogleWebfonts.js
const _ = require("lodash")
const path = require("path")
const yauzl = require("yauzl")
const fetch = require("node-fetch")
const { RawSource } = require("webpack-sources")
const FontTypes = require("./FontTypes")

const API_URL = "https://google-webfonts-helper.herokuapp.com/api/fonts"

const FONT_SRC = (font, format) => `url("${font}")${format ? ` format("${format}")` : ""}`

const FONT_FACE = ({ fontFamily, fontStyle, fontWeight, src, fallback }) => `
@font-face {
	font-family: ${fontFamily};
	font-style: ${fontStyle};
	font-weight: ${fontWeight};
	${fallback ? `src: ${fallback};` : ""}
	${src.length ? `src: ${src.join(",\n\t\t")};` : ""}
}
`

function getVariantCss({ variant, info, font, formats, fontsPath }) {
	const src = info.local.map(fileName => `local("${fileName}")`)
	let fallback
	formats.forEach(ext => {
		if(ext in info) {
			const url = (
				typeof fontsPath !== "undefined" ?
				`./${fontsPath}/${font.getName(variant)}.${ext}` :
				info[ext]
			)
			const format = FontTypes[ext]
			src.push(FONT_SRC(url, format))
			if(ext === "eot") {
				fallback = FONT_SRC(url)
			}
		}
	})
	return (
		`/* === ${font.family} - ${variant} */` +
		FONT_FACE(_.assign({ src, fallback }, info))
	)
}

class Selection {
	constructor(font, query, info) {
		this.font = font
		this.query = {
			subsets: query.subsets || font.defaults.subsets,
			variants: query.variants || font.defaults.variants,
			formats: query.formats || _.keys(FontTypes)
		}
		this.query.variants = this.query.variants.map(value => {
			switch(value) {
				case "400": return "regular"
				case "400italic": return "italic"
				default: return value
			}
		})
	}

	css(fontsPath) {
		const { font, query: { subsets, variants, formats } } = this
		return font.info(subsets)
			.then(info => {
				const css = []
				variants.forEach(variant => {
					const variantInfo = info.variants.filter(node => node.id === variant).pop()
					if(variantInfo) {
						css.push(getVariantCss({
							info: variantInfo,
							formats,
							variant,
							font,
							fontsPath
						}))
					}
				})
				return css.join("\n")
			})
	}

	download() {
		if(this._response) {
			return Promise.resolve(this._response)
		}
		const { font } = this
		const { subsets, variants, formats } = this.query
		let url = `${font.url}?download=zip`
		url += "&subsets=" + subsets.join(",")
		url += "&variants=" + variants.join(",")
		if(formats) {
			url += "&formats=" + formats.join(",")
		}
		return fetch(url)
			.then(response => {
				if(response.status !== 200) {
					throw new Error(response.statusText)
				}
				this._response = response
				return response
			})
	}

	files() {
		if(this._files) {
			return Promise.resolve(this._files)
		}
		return this.download()
			.then(response => response.buffer())
			.then(buffer => new Promise((resolve, reject) => {
				this._files = {}
				yauzl.fromBuffer(buffer, { lazyEntries: true }, (err, zipFile) => {
					if(err) {
						reject(err)
					}
					const next = () => zipFile.readEntry()
					zipFile
						.on("error", reject)
						.on("end", () => resolve(this._files))
						.on("entry", entry => {
							if(/\/$/.test(entry.fileName)) next()
							const ext = path.extname(entry.fileName).slice(1)
							const variant = entry.fileName.match(/\-([a-z0-9]+)\..*$/)[1]
							const fileName = `${this.font.getName(variant)}.${ext}`
							zipFile.openReadStream(entry, (err, stream) => {
								if(err) reject(err)
								const buffer = []
								stream.on("data", data => buffer.push(data))
								stream.on("end", () => {
									this._files[fileName] = Buffer.concat(buffer)
									next()
								})
							})
						})
					zipFile.readEntry()
				})
			}))
	}

	assets() {
		return this.files()
			.then(files => {
				const assets = {}
				for(const fileName in files) {
					assets[fileName] = new RawSource(files[fileName])
				}
				return assets
			})
	}
}

class Font {
	constructor(url, { id, family, variants, formats, defSubset, defVariant }) {
		this.apiUrl = url
		this.id = id
		this.family = family
		this.variants = variants
		this.formats = formats
		this.defaults = {
			subsets: [ defSubset ],
			variants: [ defVariant ]
		}
		this.getName = this.getName.bind(this)
	}

	get url() { return `${this.apiUrl}/${this.id}` }

	getName(variant) {
		let name = _.chain(this.family).camelCase().upperFirst().value()
		if(variant) {
			const weight = parseInt(variant.substr(0, 3))
			let suffix = "Regular"
			switch(weight) {
				case 100: suffix = "Thin"; break;
				case 200: suffix = "ExtraLight"; break;
				case 300: suffix = "Light"; break;
				case 400: suffix = "Regular"; break;
				case 500: suffix = "Medium"; break;
				case 600: suffix = "SemiBold"; break;
				case 700: suffix = "Bold"; break;
				case 800: suffix = "ExtraBold"; break;
				case 900: suffix = "Black"; break;
			}
			if(/italic$/.test(variant)) {
				suffix = (suffix === "Regular" ? "Italic" : `${suffix}Italic`)
			}
			name += `-${suffix}`
		}
		return name
	}

	info(subsets) {
		if(this._info) {
			return Promise.resolve(this._info)
		}
		let url = this.url + "?"
		if(subsets) {
			url += "subsets=" + subsets.join(",")
		}
		return fetch(url)
			.then(response => {
				if(response.status !== 200) {
					throw new Error(response.statusText)
				}
				return response.json()
			})
			.then(info => {
				this._info = info
				return info
			})
	}

	select(options = {}) {
		return new Selection(this, options)
	}
}

class GoogleWebfonts {
	constructor(apiUrl = API_URL) {
		this.url = apiUrl
	}

	getFonts() {
		if(this._fonts) {
			return Promise.resolve(this._fonts)
		} else {
			return fetch(this.url)
				.then(response => {
					if(response.status !== 200) {
						throw new Error(response.statusText)
					}
					return response.json()
				})
				.then(fonts => fonts.map(font => new Font(this.url, font)))
				.then(fonts => {
					this._fonts = fonts
					return fonts
				})
		}
	}

	getFontById(id) {
		return this.getFonts()
			.then(fonts => fonts.filter(font => font.id === id).pop())
	}

	getFontByFamily(family) {
		return this.getFonts()
			.then(fonts => fonts.filter(font => font.family.toLowerCase() === family.toLowerCase()).pop())
	}
}

GoogleWebfonts.Font = Font

GoogleWebfonts.Selection = Selection

module.exports = GoogleWebfonts
